neutron-physics/model.py

```diff
@@ -0,0 +1,387 @@
+        lines.append(f'nps {self.particles}')
+ 
+        # Write the problem
+        with open(Path('mcnp') / 'inp', 'w') as f:
+            f.write('\n'.join(lines))
+
+    def _plot(self):
+        """Extract and plot the results
+ 
+        """
+        # Read the results from the OpenMC statepoint
+        with openmc.StatePoint(Path('openmc') / 'statepoint.1.h5') as sp:
+            t = sp.get_tally(name='neutron flux')
+            x_openmc = t.find_filter(openmc.EnergyFilter).bins[:,1]*1.e-6
+            y_openmc = t.mean[:,0,0]
+ 
+        # Read the results from the MCNP output file
+        with open(Path('mcnp') / 'outp', 'r') as f:
+            text = f.read()
+            p = text.find('1tally')
+            p = text.find('energy', p) + 10
+            q = text.find('total', p)
+            t = np.fromiter(text[p:q].split(), float)
+            t.shape = (len(t) // 3, 3)
+            x_mcnp = t[1:,0]
+            y_mcnp = t[1:,1]
+            sd = t[1:,2]
+ 
+        # Normalize the spectra
+        y_openmc /= np.diff(np.insert(x_openmc, 0, 1.e-11))*sum(y_openmc)
+        y_mcnp /= np.diff(np.insert(x_mcnp, 0, 1.e-11))*sum(y_mcnp)
+ 
+        # Compute the relative error
+        err = np.zeros_like(y_mcnp)
+        idx = np.where(y_mcnp > 0)
+        err[idx] = (y_openmc[idx] - y_mcnp[idx])/y_mcnp[idx]
+ 
+        # Set up the figure
+        fig = plt.figure(1, facecolor='w', figsize=(8,8))
+        ax1 = fig.add_subplot(111)
+ 
+        # Create a second y-axis that shares the same x-axis, keeping the first
+        # axis in front
+        ax2 = ax1.twinx()
+        ax1.set_zorder(ax2.get_zorder() + 1)
+        ax1.patch.set_visible(False)
+ 
+        # Plot the spectra
+        ax1.loglog(x_mcnp, y_mcnp, 'r', linewidth=1, label='MCNP')
+        ax1.loglog(x_openmc, y_openmc, 'b', linewidth=1, label='OpenMC', linestyle='--')
+ 
+        # Plot the relative error and uncertainties
+        ax2.semilogx(x_mcnp, err, color=(0.2, 0.8, 0.0), linewidth=1)
+        ax2.semilogx(x_mcnp, 2*sd, color='k', linestyle='--', linewidth=1)
+        ax2.semilogx(x_mcnp, -2*sd, color='k', linestyle='--', linewidth=1)
+ 
+        # Set grid and tick marks
+        ax1.tick_params(axis='both', which='both', direction='in', length=10)
+        ax1.grid(b=False, axis='both', which='both')
+        ax2.tick_params(axis='y', which='both', right=False)
+        ax2.grid(b=True, which='both', axis='both', alpha=0.5, linestyle='--')
+ 
+        # Set axes labels and limits
+        ax1.set_xlim([1.e-11, self.energy_mev])
+        ax1.set_xlabel('Energy (MeV)', size=12)
+        ax1.set_ylabel('Spectrum', size=12)
+        ax1.legend()
+        ax2.set_ylabel("Relative error", size=12)
+        title = f'{self.nuclide}, {self.energy_mev} MeV Source'
+        plt.title(title)
+ 
+        # Save plot
+        os.makedirs('plots', exist_ok=True)
+        if self.name is None:
+            name = f'{self.nuclide}.png'
```

Comment:
Also include energy in the filename as we do for photons? Maybe temperature too?

Reply:
Yeah, I'll make that the default.